### PR TITLE
Updating to stac pydantic 3

### PR DESCRIPTION
**Pull request by rhysrevans3**
**Related Issue(s):**

- #

**Description:**

**PR Checklist:**

- [x] `pre-commit` hooks pass locally
- [x] Tests pass (run `make test`)
- [x] Documentation has been updated to reflect changes, if applicable, and docs build successfully (run `make docs`)
- [x] Changes are added to the [CHANGELOG](https://github.com/stac-utils/stac-fastapi/blob/main/CHANGES.md).

### CODE REVIEW

**Comment by thomas-maschler** [comment]
@rhysrevans3 looks like we had the same idea
https://github.com/stac-utils/stac-fastapi/pull/625

File: stac_fastapi/api/stac_fastapi/api/models.py [hunk 1]
```diff
@@ -44,31 +43,8 @@ def create_request_model(
...
-                    min_items=field_info.min_items,
-                    max_items=field_info.max_items,
-                    min_length=field_info.min_length,
-                    max_length=field_info.max_length,
-                    regex=field_info.regex,
-                    extra=field_info.extra,
-                )
-                fields[k] = (v.outer_type_, body)
+            for k, field_info in model.model_fields.items():
+                fields[k] = (field_info.annotation, field_info)
```
Comment (rhysrevans3) [inline stac_fastapi/api/stac_fastapi/api/models.py:47]: @thomas-maschler sorry I should have checked before putting in the pull request it looks like most of our changes are the same. I found the `field_info.annotation` would leave out the `Optional` type unless I switched from typing List and Dict to the inbuilt list and dict.

File: CHANGES.md [hunk 2]
```diff
@@ -2,6 +2,9 @@
 
 ## [Unreleased]
 
+* Removing support for Python 3.8
```
Comment (vincentsarago) [inline CHANGES.md:5]: Why is this needed? pydantic 2 does support python 3.8 https://github.com/pydantic/pydantic/blob/main/pyproject.toml#L65

File: Dockerfile [hunk 3]
```diff
@@ -1,4 +1,4 @@
-FROM python:3.8-slim as base
+FROM python:3.9-slim as base
```
Comment (vincentsarago) [inline Dockerfile:1]: agree to remove python 3.8 but then we should use python 3.11 IMO

File: stac_fastapi/api/stac_fastapi/api/app.py [hunk 4]
```diff
@@ -67,8 +68,8 @@ class StacApi:
 
     settings: ApiSettings = attr.ib()
     client: Union[AsyncBaseCoreClient, BaseCoreClient] = attr.ib()
-    extensions: List[ApiExtension] = attr.ib(default=attr.Factory(list))
-    exceptions: Dict[Type[Exception], int] = attr.ib(
+    extensions: list[ApiExtension] = attr.ib(default=attr.Factory(list))
+    exceptions: dict[Type[Exception], int] = attr.ib(
```
Comment (vincentsarago) [inline stac_fastapi/api/stac_fastapi/api/app.py:72]: FYI is using `dict` directly instead of `typing.Dict` is the reason to switch to python>=3.9, `typing.Dict` is just an alias of `dict` (https://github.com/python/cpython/blob/3.9/Lib/typing.py#L1743C15-L1743C19) so it's fine to use `typing.Dict` IMO 

https://stackoverflow.com/questions/37087457/difference-between-defining-typing-dict-and-dict

Comment (vincentsarago) [reply stac_fastapi/api/stac_fastapi/api/app.py:72]: same goes for List

Comment (rhysrevans3) [reply stac_fastapi/api/stac_fastapi/api/app.py:72]: @vincentsarago I found that during the `create_request_model` in models.py. The `field_info.annotation` would leave out the `Optional` part of the type. The only solution I found was to switch from typing List and Dict to the generic list and dict. There maybe a better solution to this?

**Comment by lossyrob** [comment]
Is there a reason to keep both this and #625 open? Based on a quick review, looks like #625 has less changes - and by the author of the stac-pydantic upgrade (thank you @thomas-maschler !). If we can close this and focus the review on #625 that will be easier on reviewers. @rhysrevans3 anything in here related to the pydantic upgrade that deem a must-have that #625 doesn't already contain? Otherwise consolidating on #625 and getting your review on it would be a good path to get this unblocked IMO. Thanks for this PR, unfortunate instance of duplication of work!
